miniapps/navier/navier_particles.cpp

```diff
@@ -0,0 +1,457 @@
+         p_xn = p_xc;
+
+         // Set order to 0, to avoid re-computing x history as well
+         Order()[i] = 0;
+      }
+
+   }
+}
+
+void NavierParticles::ApplyBCs()
+{
+   for (BCVariant &bc_v : bcs)
+   {
+      std::visit(
+         [this](auto &bc)
+      {
+         using T = std::decay_t<decltype(bc)>;
+         if constexpr(std::is_same_v<T, ReflectionBC_2D>)
+         {
+            Apply2DReflectionBC(bc);
+         }
+         else if constexpr(std::is_same_v<T, RecirculationBC_2D>)
+         {
+            Apply2DRecirculationBC(bc);
+         }
+      },bc_v);
+
+   }
+}
+
+NavierParticles::NavierParticles(MPI_Comm comm, int num_particles, Mesh &m)
+   : fluid_particles(comm, num_particles, m.SpaceDimension()),
+     inactive_fluid_particles(comm, 0, m.SpaceDimension()),
+     finder(comm)
+{
+
+   for (int o = 0; o < 3; o++)
+   {
+      beta_k[o].SetSize(4);
+      alpha_k[o].SetSize(3);
+   }
+
+   finder.Setup(m);
+
+   int dim = fluid_particles.GetDim();
+
+   // Initialize kappa, zeta, gamma
+   fp_idx.field.kappa = fluid_particles.AddField(1, Ordering::byVDIM, "kappa");
+   fp_idx.field.zeta = fluid_particles.AddField(1, Ordering::byVDIM, "zeta");
+   fp_idx.field.gamma = fluid_particles.AddField(1, Ordering::byVDIM, "gamma");
+
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "kappa");
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "zeta");
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "gamma");
+
+   // Initialize fluid particle fields
+   for (int i = 0; i < 4; i++)
+   {
+      string suffix = i > 0 ? "_nm" + to_string(i) : "_n";
+      fp_idx.field.u[i] = fluid_particles.AddField(dim, Ordering::byVDIM,
+                                                   ("u" + suffix).c_str());
+      fp_idx.field.v[i] = fluid_particles.AddField(dim, Ordering::byVDIM,
+                                                   ("v" + suffix).c_str());
+      fp_idx.field.w[i] = fluid_particles.AddField(dim, Ordering::byVDIM,
+                                                   ("w" + suffix).c_str());
+      if (i > 0)
+      {
+         fp_idx.field.x[i-1] = fluid_particles.AddField(dim, Ordering::byVDIM,
+                                                        ("x" + suffix).c_str());
+      }
+   }
+
+   // Initialize order (tag)
+   fp_idx.tag.order = fluid_particles.AddTag("Order");
+   fluid_particles.Tag(fp_idx.tag.order) = 0;
+
+   // Reserve num_particles for inactive_fluid_particles
+   inactive_fluid_particles.Reserve(num_particles);
+
+   r.SetSize(dim);
+   C.SetSize(dim);
+}
+
+void NavierParticles::Step(const real_t &dt, const ParGridFunction &u_gf,
+                           const ParGridFunction &w_gf)
+{
+   // Shift fluid velocity, fluid vorticity, particle velocity, and particle position
+   for (int i = 3; i > 0; i--)
+   {
+      U(i) = U(i-1).GetData();
+      V(i) = V(i-1).GetData();
+      W(i) = W(i-1).GetData();
+      X(i) = X(i-1).GetData();
+   }
```

Comment:
This is horribly inefficient 😮 Just do it in ring buffer fashion keeping index of the head and length. Then increment the index of the head here and increase the length for the first few steps before it gets full, so that is like two integer operations instead of copying GBs of data potentially 😨 .

Reply:
Keeping this for now since it is readable and this is a 2D example right now. We can optimize it when we add the 3D extension later this year.

Reply:
Yep, but it is equally important that a new user can easily understand this code. Also, we will be extending this miniapp to 3D and GPUs so the code will evolve regardless of what we do now.